groups/bsl/bslstl/bslstl_optional.h

```diff
@@ -704,55 +1010,415 @@ struct Optional_Data<TYPE, true> : public Optional_DataImp<TYPE> {
+    {
+        // Must be in-place inline because the use of 'enable_if' will
+        // otherwise break the MSVC 2010 compiler.
+        this->emplace(BSLS_COMPILERFEATURES_FORWARD(ANY_TYPE, rhs));
+    }
+
+    template <class ANY_TYPE>
+    optional(const optional<ANY_TYPE>& rhs
+        BSLSTL_OPTIONAL_ENABLE_IF_CONSTRUCT_FROM_OPTIONAL_LVAL
+        BSLSTL_OPTIONAL_ENABLE_IF_NOT_EXPLICIT_CONSTRUCT(TYPE,
+            const ANY_TYPE&))
+        // If specified 'rhs' contains a value, initialize the 'value_type'
+        // object using '*rhs'.  Otherwise, create a disengaged optional.  Use
+        // the currently installed default allocator to supply memory for this
+        // and any future 'value_type' objects.
+    {
+        // Must be in-place inline because the use of 'enable_if' will
+        // otherwise break the MSVC 2010 compiler.
+        if (rhs.has_value()) {
+            emplace(rhs.value());
+        }
+    }
+
+    template <class ANY_TYPE>
+    explicit optional(const optional<ANY_TYPE>& rhs
+        BSLSTL_OPTIONAL_ENABLE_IF_CONSTRUCT_FROM_OPTIONAL_LVAL
+        BSLSTL_OPTIONAL_ENABLE_IF_EXPLICIT_CONSTRUCT(TYPE, const ANY_TYPE&))
+        // If specified 'rhs' contains a value, initialize the 'value_type'
+        // object using '*rhs'.  Otherwise, create a disengaged optional.  Use
+        // the currently installed default allocator to supply memory for this
+        // and any future 'value_type' objects.
+    {
+        // Must be in-place inline because the use of 'enable_if' will
+        // otherwise break the MSVC 2010 compiler.
+        if (rhs.has_value()) {
+            emplace(rhs.value());
+        }
+    }
+
+#if defined(BSLS_COMPILERFEATURES_SUPPORT_RVALUE_REFERENCES)
+    // 'MovableRef' prevents correct type deduction in C++11 when used with
+    // 'optional<ANY_TYPE>'.  These constructors need to be defined in terms of
+    // rvalue reference in C++11.  In C++03, this type deduction issue does not
+    // exist due to the nature of C++03 MovableRef implementation and usage.
+    // Consequently, a 'MovableRef' equivalent constructors needs to be
+    // provided in C++03 (see below).
+    template <class ANY_TYPE>
+    optional(optional<ANY_TYPE>&& rhs
+        BSLSTL_OPTIONAL_ENABLE_IF_CONSTRUCT_FROM_OPTIONAL_RVAL
+        BSLSTL_OPTIONAL_ENABLE_IF_CONSTRUCT_DOES_NOT_PROPAGATE_ALLOCATOR
+        BSLSTL_OPTIONAL_ENABLE_IF_NOT_EXPLICIT_CONSTRUCT(TYPE, ANY_TYPE))
+        // If specified 'rhs' contains a value, initialize the 'value_type'
+        // object by moving from '*rhs'.  Otherwise, create a disengaged
+        // 'optional'.  Use the currently installed default allocator to supply
+        // memory for this and any future 'value_type' objects.
+    {
+        // Must be in-place inline because the use of 'enable_if' will
+        // otherwise break the MSVC 2010 compiler.
+        if (rhs.has_value()) {
+            emplace(MoveUtil::move(rhs.value()));
```

Comment:
```suggestion
            emplace(std::move(rhs.value()));
```
We're in c++11+ land.  Using `std::move` is better.

Reply:
done. applied throughout the code.